Feat/easy swapper v2 withdrawals

dimlbc opened the pull request:
https://trello.com/c/yQ2eaT7w/4599-easyswapper-v2-withdrawals

**PR is ready to be reviewed, unit tests are in progress and will be fixed and added soon**

Updated single asset withdraw flow to use new easy swapper v2.

Withdraw flow:

**1. User can select between to options:**

1a. Multi asset withdraw (using PoolLogic contract)

<img width="669" alt="Screenshot 2024-10-14 at 18 32 07" src="https://github.com/user-attachments/assets/f92e831a-25e3-499a-9c87-42de338bc896">
<img width="633" alt="Screenshot 2024-10-14 at 18 26 13" src="https://github.com/user-attachments/assets/ef8000b4-99c6-4b7e-9013-12e2318f2f39">

**1b. Single asset withdraw**

<img width="677" alt="Screenshot 2024-10-14 at 18 25 57" src="https://github.com/user-attachments/assets/22a624eb-992d-44b4-be0f-cf0888d0819d">

2. To withdraw into a single asset, a user now needs to execute two transactions. After the first unroll transaction is executed, the user will see the received assets inside the withdrawal contract and will have two options: swap into a single asset or claim without swapping.

<img width="671" alt="Screenshot 2024-10-14 at 18 27 51" src="https://github.com/user-attachments/assets/e54c56f1-2771-4c62-99f0-8850fffd0883">

<img width="646" alt="Screenshot 2024-10-14 at 18 28 42" src="https://github.com/user-attachments/assets/e5fa63ac-971b-4265-8345-26e2ac21afca">
<img width="642" alt="Screenshot 2024-10-14 at 18 28 56" src="https://github.com/user-attachments/assets/ffca3bee-eaf9-477b-88a0-9bf03d5e6c4e">



### PR can be tested locally on Base.

1. Add working alchemy api key in packages/trading-widget/src/core-kit/providers/wagmi-provider.tsx
2. Implement `getSwapData` callback in packages/trading-widget/src/examples/simple-example.tsx
3. `pnpm storybook`

artsiomYavorski commented on packages/trading-widget/src/core-kit/hooks/trading/withdraw-v2/use-min-withdraw-slippage.ts:
```diff
@@ -0,0 +1,49 @@
+  const { defaultWithdrawSlippage, defaultSwapTransactionSlippage } =
+    useConfigContextParams()
+  const isWithdraw = !useIsDepositTradingPanelType()
+  const isWithdrawSwapStep = useIsWithdrawSwapStep()
```
Would it be helpful  to return  also `isFetching`  state from `useIsWithdrawSwapStep`  to make effect more stable from flicks?

dimlbc replied:
I added prefetch of `trackedAssets` on deposit tab to prevent flicks, but will also add `isFetching` state to be on a safe side

artsiomYavorski commented:
@dimlbc wdyt of renaming `skipSwap` flag to `isClaim`?

artsiomYavorski commented on packages/trading-widget/src/core-kit/hooks/trading/withdraw-v2/swap-step/use-withdraw-swap-data.ts:
```diff
@@ -0,0 +1,45 @@
+  const { data: assets = [] } = useWithdrawTrackedAssets()
+  const swapDataRequired =
+    assets.length > 0 &&
+    !assets.every(({ address }) =>
```
```suggestion
    assets.some(({ address }) =>
```

artsiomYavorski commented on packages/trading-widget/src/core-kit/hooks/trading/withdraw-v2/swap-step/use-withdraw-swap-quote.ts:
```diff
@@ -0,0 +1,48 @@
+  }, [isFetching, updateReceiveToken])
+
+  useEffect(() => {
+    if (assets.length === 0 || isFetching || isError) {
```
Is there a chance where `assets.length === 0` could be a valid value and should be processed?

dimlbc replied:
No, total received value can be calculated only when assets.length > 0

artsiomYavorski commented on packages/trading-widget/src/core-kit/hooks/trading/withdraw-v2/swap-step/use-withdraw-swap-transaction.ts:
```diff
@@ -0,0 +1,95 @@
+    })
+
+    console.log('Function', functionName)
+    console.log('Arguments', txArgs)
```
Having it intentionally?

dimlbc replied:
Yes, instead of previous logging of transaction arguments

artsiomYavorski commented on packages/trading-widget/src/trading-widget/components/widget/widget-overlay/trading-overlay/trading-summary/claim-summary.tsx:
```diff
@@ -0,0 +1,28 @@
+
+  return (
+    <div>
+      <span>Claim</span>
```
Please use translation context

artsiomYavorski commented on packages/trading-widget/src/trading-widget/components/widget/widget-overlay/trading-overlay/trading-summary/swap-summary.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  return (
+    <div className="dtw-flex dtw-flex-wrap dtw-items-center dtw-justify-center dtw-gap-1">
+      Swap of{' '}
```
use translation key

artsiomYavorski commented on packages/trading-widget/src/trading-widget/components/withdraw/swap-step/input-group/input-group.hooks.ts:
```diff
@@ -0,0 +1,24 @@
+  const [receiveToken] = useReceiveTokenInput()
+  const price = useAssetPrice({ address: receiveToken.address, chainId })
+  const swapDiff = useSwapReceiveValueDiff()
+  const themeType = useGetThemeTypeBySlippage(swapDiff < 0 ? swapDiff : 0)
```
```suggestion
  const themeType = useGetThemeTypeBySlippage(Math.max(0, swapDiff))
```

artsiomYavorski commented on packages/trading-widget/src/trading-widget/components/withdraw/unroll-step/input-group/withdraw-section/widget-section.hooks.ts:
```diff
@@ -0,0 +1,20 @@
+  assetSymbol,
+}: WithdrawSectionProps) => {
+  const [sendToken] = useSendTokenInput()
+  const label = isMultiAssetWithdraw ? 'Receive(estimated)' : ''
```
translation key

artsiomYavorski commented on packages/trading-widget/src/trading-widget/components/withdraw/unroll-step/input-group/withdraw-section/withdraw-section.tsx:
```diff
@@ -0,0 +1,45 @@
+              </div>
+              <ul>
+                <li className="dtw-mt-1 dtw-flex dtw-gap-1">
+                  1. Unroll <TokenIcon symbols={[vaultSymbol]} size="sm" />{' '}
```
translation keys in list

dimlbc commented on packages/trading-widget/src/core-kit/hooks/trading/use-swaps-data-query.ts:
```diff
@@ -0,0 +1,44 @@
+  return useQuery({
+    queryKey: ['getSwapsData', assets],
+    queryFn: async ({ signal, queryKey: [, assets] }) =>
+      Promise.all(
```
Partial data is not relevant here, so we don't need to use Promise.allSettled here

dimlbc commented:
Added stepper and updated steps description

<img width="388" alt="Screenshot 2024-10-15 at 13 46 33" src="https://github.com/user-attachments/assets/9084fceb-bd13-4bc8-b5a7-6f67e1b89e91">
<img width="394" alt="Screenshot 2024-10-15 at 13 46 54" src="https://github.com/user-attachments/assets/aa2072c7-0b77-4174-a12a-e19c04bf16a8">

D-Ig commented on packages/trading-widget/src/trading-widget/providers/translation-provider/translation-provider.defaults.ts:
```diff
@@ -98,4 +98,16 @@ export const DEFAULT_TRANSLATION_DATA: TranslationMap = {
+  swapOf: 'Swap of',
+  to: 'to',
+  initWithdrawDescription:
+    'Unroll {vaultSymbol} tokens into convertible tokens',
```
```suggestion
    'Unroll {vaultSymbol} tokens into swappable assets',
```

D-Ig commented on packages/trading-widget/src/trading-widget/providers/translation-provider/translation-provider.defaults.ts:
```diff
@@ -98,4 +98,16 @@ export const DEFAULT_TRANSLATION_DATA: TranslationMap = {
+  initWithdrawDescription:
+    'Unroll {vaultSymbol} tokens into convertible tokens',
+  initWithdrawTooltip:
+    'Convertible tokens are basic ERC20 tokens which can be swapped on any DEXs',
```
```suggestion
    'Swappable assets are basic ERC20 tokens which can be converted on various platforms, including decentralized exchanges (DEXs), swap aggregators, and centralized exchanges (CEXs).',
```

D-Ig reviewed (commented):
what will happen if offchain swap data is not available? I reckon a UX-friendly flow should be developed (if not already), which will guide the user to withdraw using 2nd option. Same can be applied to deposits (if zap deposit fail -> suggest to deposit w/o swap data).

this can go separately, I think

dimlbc commented:
> what will happen if offchain swap data is not available? I reckon a UX-friendly flow should be developed (if not already), which will guide the user to withdraw using 2nd option. Same can be applied to deposits (if zap deposit fail -> suggest to deposit w/o swap data).
> 
> this can go separately, I think

Sure, we need to implement it. Let's do in separate scope

dimlbc commented on packages/trading-widget/src/core-kit/hooks/trading/withdraw-v2/complete-step/use-complete-withdraw-transaction.ts:
```diff
@@ -0,0 +1,97 @@
+    })
+
+    console.log('Function', functionName)
+    console.log('Arguments', txArgs)
```
removed logging of `Function` here and during init withdraw

dimlbc commented on packages/trading-widget/src/core-kit/hooks/trading/withdraw-v2/complete-step/use-complete-withdraw-transaction.ts:
```diff
@@ -0,0 +1,97 @@
+      chainId: poolConfig.chainId,
+    })
+
+    console.log('Function', functionName)
```
removed logging
